### ISSUE TITLE
Create a TwitterDescription component

### ISSUE THREAD

**Issue by andizer** [Yoast/javascript#49]
The `TwitterDescription` is shown as a paragraph in the Twitter preview:
<img width="526" alt="yoast___yoast____twitter" src="https://user-images.githubusercontent.com/325040/49155171-8c19cb00-f31a-11e8-807d-a9ae15ad092f.png">

## Properties
* description (maximum length of 200 characters).

## Styling
The component should have the following CSS props:
```css
    font-size: 14px;
    max-height: 36.4px;
    line-height: 18.2px;
    overflow: hidden;
    margin-bottom: 4.52662px;
    width: 476px;
```

When the card type `Summary with large image` the paragraph will have a `max-height` of 36.4px. 

When a squared image (`summary` card or image has ratio of 1:1) has been used the `width` will be 357.5px and the `max-height` will be 54.6px
<img width="525" alt="_3__twitter" src="https://user-images.githubusercontent.com/325040/49155937-70afbf80-f31c-11e8-8629-6ff7c8375ba8.png">

Parent https://github.com/Yoast/javascript/issues/51

**Comment by hwinne** [comment]
@andizer With the current css styling, it is not possible to display 200 chars. For example, the Twitter preview as pasted above contains a description of approximately 157 characters, even though 200 is the max and the description obviously is not rendered completely.

Should the height be changed, or is 2 lines the maximum?

**Comment by hwinne** [comment]
@andizer We have another question. In the preview example, we can see that the description text spans two lines, and ends in ellipsis. However, the CSS docs explain that `white-space: nowrap` is required for use of `text-overflow: ellipsis`:

<img width="200" alt="Screenshot 2019-03-22 at 11 33 45" src="https://user-images.githubusercontent.com/42736463/54817375-0dfd7d80-4c97-11e9-9ab6-c7d734b24536.png">

<img width="400" alt="Screenshot 2019-03-22 at 11 34 02" src="https://user-images.githubusercontent.com/42736463/54817384-135ac800-4c97-11e9-8fe4-7d1ce2cf79e7.png">

By definition, `white-space: nowrap` will never wrap text to the next line.

So the question is: how is it possible that the screenshot has text wrapped over two lines, and also has ellipsis (`...`)?

**Comment by johannadevos** [comment]
Note: Before we can look into @andizer's remarks regarding the CSS props when there is an image, the TwitterImage component needs to be created. @hwinne is working on that at the moment.

**Comment by johannadevos** [comment]
@afercia Your help would be very welcome :)

**Comment by afercia** [comment]
Correct: `text-overflow: ellipsis` works only with `white-space: nowrap` so it applies only to single lines. There's no way to do this in CSS other than using proprietary CSS properties, not supported in all browsers.

If you look at the markup of the example above you will notice the text is actually truncated after a certain amount of characters before it's sent to the browser: it's not done via CSS.

<img width="703" alt="Screenshot 2019-03-25 at 10 26 07" src="https://user-images.githubusercontent.com/1682452/54908577-85761b80-4ee8-11e9-9062-25b9973812b9.png">

**Comment by johannadevos** [comment]
@afercia Thank you for the explanation!

It seems that the truncation issue is more complicated than foreseen. If we cannot do it in CSS, what is the best approach? We can manually implement truncation (see [here](https://gomakethings.com/how-to-truncate-text-with-vanilla-javascript/) for an example), but that tutorial isn't completely applicable to our situation since we do not want to truncate after a certain number of words. In our case, we want to fill the available space (two lines with a certain width) with characters, and then truncate.

**Comment by afercia** [comment]
There's no perfect solution, I guess.

It is possible to truncate after a certain amount of characters. I think @igorschoester implemented something like that for the TwitterTitle.

However, truncating at a fixed amount of characters (say 200) won't produce the same results in terms of _space_: for example: 200 `i` take less space than 200 `m`.

Re: CSS there's `line-clamp` but it's not standard and only works in webkit (Chrome, Safari, Opera). See https://css-tricks.com/almanac/properties/l/line-clamp/

In the same article, they mention [Clamp.js](https://github.com/josephschmitt/Clamp.js) haven't tested it, seems a bit old. Also, not sure using JS for this is appropriate, I wouldn't recommend it. Anyways, adopting new npm packages should be evaluated by @atimmer :)

**Comment by atimmer** [comment]
@johannadevos @afercia We have solved this in a different way for the snippet preview: https://github.com/Yoast/javascript/blob/develop/packages/search-metadata-previews/src/snippet-preview/SnippetPreview.js#L331-L357. We resize the title until it is small enough to fit on one line (two lines on mobile I think). When it was at first too large then we render `...` behind it. This is a way to approximate what Google/Twitter do.

Let's use the same method for the TwitterPreview. Because this is now duplicated, please abstract this into a generally useable higher order component that lives in `@yoast/helpers`. You could also split this resizing logic from this issue and do it later, but I leave that to @IreneStr.

**Comment by IreneStr** [comment]
I discussed the description length a bit in the #product channel, because the Twitter docs are inconsistent, and not reflecting the reality in some regards. The conclusion is that we should go for:
- `summary`: 3 lines maximum. The width of each line is 476px.
- `summary-large-image`: 2 lines maximum. The width of each line is 476px.

I'll create a new issue for the resizing, because it's also needed for the already merged TwitterTitle component.
